### PR TITLE
MGDOBR-305: Test Webhook action

### PR DESCRIPTION
**Pull request by jomarko**
For more details see:
- https://issues.redhat.com/browse/MGDOBR-305

## Related PRs
- https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/507

## PR owner checklist

Please make sure that your PR meets the following requirements:

- [ ] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/main/kogito-build/kogito-ide-config).  
  *Run `mvn clean verify -DskipTests` so that imports are correctly set.*
- [ ] Your commit messages are clear and reference the JIRA issue e.g: "[MGDOBR-1] - $clear_explanation_of_what_you_did"
- [ ] All new functionality is tested
- [ ] Pull Request title is properly formatted: `MGDOBR-XYZ Subject`
- [ ] Pull Request contains link to the JIRA issue
- [ ] Pull Request contains link to any dependent or related Pull Request
- [ ] Pull Request contains description of the issue
- [ ] Pull Request does not include fixes for issues other than the main ticket

## PR reviewer(s) checklist

- [ ] Check if code and Github action workflows have been modified and if the modification is safe
- [ ] If the modification is safe, add the `safe to test` label

<details>
<summary>
How to trigger pipelines and use the bots:
</summary>

- <b>Run the end to end pipeline</b>  
  Annotate the pull request with the label: `safe to test`. If you want to run the pipeline again, remove and add it again.

- <b>Rebase the pull request</b>  
  Comment with: `/rebase`.

</details>

### CODE REVIEW

File: integration-tests/src/test/java/com/redhat/service/bridge/integration/tests/steps/WebhookSiteSteps.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    //    private static final String ENDPOINT_UUID = "a0704e8f-a817-4d02-b30a-b8c49d0132dc";
+    private static final String ENDPOINT_UUID = Utils.getSystemProperty("webhook.site.uuid");
+
+    private TestContext context;
+
+    public WebhookSiteSteps(TestContext context) {
+        this.context = context;
+    }
+
+    @And("^Webhook site request history is cleared$")
```
Comment (sutaakar) [inline integration-tests/src/test/java/com/redhat/service/bridge/integration/tests/steps/WebhookSiteSteps.java:32]: This step can cause a race condition (i.e. running 2 PR checks at the same time).
One possibility could be to rather implement a hook checking all requests from site and deleting all requests older than ~24 hours, what do you think?

File: integration-tests/src/test/resources/features/webhookaction.feature [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    Then Webhook site contains request with text "hello world by ${bridge.mybridge.cloud-event.webhook-test.id}" within 1 minute
+
+
+    When delete the Processor "myProcessor" of the Bridge "mybridge"
+    
+    Then the Processor "myProcessor" of the Bridge "mybridge" is not existing within 2 minutes
+
+    And the Manager metric 'manager_processor_status_change_total{status="PROVISIONING",}' count is at least 1
+    And the Manager metric 'manager_processor_status_change_total{status="READY",}' count is at least 1
+    And the Manager metric 'manager_processor_status_change_total{status="DELETED",}' count is at least 1
```
Comment (sutaakar) [inline integration-tests/src/test/resources/features/webhookaction.feature:51]: These metrics are already checked on https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/blob/main/integration-tests/src/test/resources/features/processor.feature so I would remove it from here as they are unrelated to the WebHook test.

File: integration-tests/src/test/resources/features/webhookaction.feature [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        }
+    }
+    """
+    Then the Ingress of the Bridge "mybridge" metric 'http_server_requests_seconds_count{method="POST",outcome="SUCCESS",status="200",uri="/events",}' count is at least 1
+    Then Webhook site contains request with text "hello world by ${bridge.mybridge.cloud-event.webhook-test.id}" within 1 minute
+
+
+    When delete the Processor "myProcessor" of the Bridge "mybridge"
+    
+    Then the Processor "myProcessor" of the Bridge "mybridge" is not existing within 2 minutes
```
Comment (sutaakar) [inline integration-tests/src/test/resources/features/webhookaction.feature:47]: Not needed as this is handled by cleanup script

File: integration-tests/src/test/resources/features/webhookaction.feature (hunk 2 above)
Comment (jomarko) [reply integration-tests/src/test/resources/features/webhookaction.feature]: @sutaakar sure, will do

File: integration-tests/src/test/resources/features/webhookaction.feature (hunk 3 above)
Comment (jomarko) [reply integration-tests/src/test/resources/features/webhookaction.feature]: sure, I will do @sutaakar

File: integration-tests/src/test/java/com/redhat/service/bridge/integration/tests/steps/WebhookSiteSteps.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+                .statusCode(204);
+    }
+
+    @Then("^Webhook site contains request with text \"([^\"]*)\" within (\\d+) (?:minute|minutes)$")
+    public void webhookSiteContainsRequest(String messageText, int timeoutMinutes) {
+        String messageTextWithoutPlaceholders = ContextResolver.resolveWithScenarioContext(context, messageText);
+        Awaitility.await()
+                .atMost(Duration.ofMinutes(timeoutMinutes))
+                .pollInterval(Duration.ofSeconds(1))
+                .untilAsserted(() -> assertThat(RestAssured.get(ENDPOINT_BASE_URL + "token/" + ENDPOINT_UUID + "/requests")
```
Comment (sutaakar) [inline integration-tests/src/test/java/com/redhat/service/bridge/integration/tests/steps/WebhookSiteSteps.java:46]:
```suggestion
                .untilAsserted(() -> assertThat(RestAssured.get(ENDPOINT_BASE_URL + "token/{webhookUuid}/requests", ENDPOINT_UUID)
```

File: integration-tests/src/test/resources/features/webhookaction.feature [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    And Webhook site request history is cleared
+
+    And add a Processor to the Bridge "mybridge" with body:
+    """
+    {
+      "name": "myProcessor",
+      "action": {
+        "type": "Webhook",
+        "parameters": {
+            "endpoint": "https://webhook.site/${system.webhook.site.uuid}"
```
Comment (sutaakar) [inline integration-tests/src/test/resources/features/webhookaction.feature:18]:
```suggestion
            "endpoint": "https://webhook.site/${env.webhook.site.uuid}"
```

File: integration-tests/src/test/resources/features/webhookaction.feature [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        },
+        "transformationTemplate" : "{\"text\": \"hello {data.name} by {id}\"}"
+      }
+    }
+    """
+    And the list of Processor instances of the Bridge "mybridge" is containing the Processor "myProcessor"
+    And the Processor "myProcessor" of the Bridge "mybridge" is existing with status "ready" within 3 minutes
+    And the Processor "myProcessor" of the Bridge "mybridge" has action of type "Webhook" and parameters:
+      | endpoint | https://webhook.site/${system.webhook.site.uuid} |
```
Comment (sutaakar) [inline integration-tests/src/test/resources/features/webhookaction.feature:27]:
```suggestion
      | endpoint | https://webhook.site/${env.webhook.site.uuid} |
```

File: integration-tests/src/test/resources/features/webhookaction.feature [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+      "specversion": "1.0",
+      "type": "webhook.site.invoked",
+      "source": "WebhookActionTestService",
+      "id": "webhook-test",
+      "data": {
+          "name": "world"
+        }
+    }
+    """
+    Then the Ingress of the Bridge "mybridge" metric 'http_server_requests_seconds_count{method="POST",outcome="SUCCESS",status="200",uri="/events",}' count is at least 1
```
Comment (sutaakar) [inline integration-tests/src/test/resources/features/webhookaction.feature:41]: This line is not needed.

File: integration-tests/src/test/resources/features/webhookaction.feature [hunk 8]
```diff
@@ -0,0 +1,42 @@
+Feature: Webhook Action tests
```
Comment (sutaakar) [inline integration-tests/src/test/resources/features/webhookaction.feature:1]: Please rename the file to `action-webhook.feature` to keep the consistency.

File: integration-tests/src/test/resources/features/webhookaction.feature [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+  @webhookaction
+  Scenario: Webhook is correctly called
+    Given authenticate against Manager
+    And create a new Bridge "mybridge"
+    And the Bridge "mybridge" is existing with status "ready" within 2 minutes
```
Comment (sutaakar) [inline integration-tests/src/test/resources/features/webhookaction.feature:7]:
```suggestion
    And the Bridge "mybridge" is existing with status "ready" within 4 minutes
```

File: integration-tests/src/test/java/com/redhat/service/bridge/integration/tests/steps/WebhookSiteSteps.java (hunk 1 above)
Comment (jomarko) [reply integration-tests/src/test/java/com/redhat/service/bridge/integration/tests/steps/WebhookSiteSteps.java:30]: @sutaakar I tried to implement your suggestion as https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/494/commits/d315c82743c7a66aab9d093ab85efa0b3796c10f , please let me know your opinion?

File: integration-tests/src/test/java/com/redhat/service/bridge/integration/tests/steps/WebhookSiteSteps.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+                                                        .then()
+                                                        .extract()
+                                                        .body()
+                                                        .jsonPath().getList("data.content")).contains(messageText));
```
[the remark this answers is not in the document]
Comment (jomarko) [reply integration-tests/src/test/java/com/redhat/service/bridge/integration/tests/steps/WebhookSiteSteps.java]: @sutaakar good point, please check in latest version

File: integration-tests/src/test/resources/features/webhookaction.feature (hunk 7 above)
Comment (jomarko) [reply integration-tests/src/test/resources/features/webhookaction.feature]: ok, will remove it

File: integration-tests/src/test/resources/features/webhookaction.feature (hunk 8 above)
Comment (jomarko) [reply integration-tests/src/test/resources/features/webhookaction.feature]: sure

File: integration-tests/src/test/resources/features/webhookaction.feature (hunk 9 above)
Comment (jomarko) [reply integration-tests/src/test/resources/features/webhookaction.feature]: @sutaakar could you please clarify, I see 2 minutes also in action-slack.feature. Should I increase the timeout also there?

File: integration-tests/src/test/java/com/redhat/service/bridge/integration/tests/steps/WebhookSiteSteps.java (hunk 1 above)
Comment (sutaakar) [reply integration-tests/src/test/java/com/redhat/service/bridge/integration/tests/steps/WebhookSiteSteps.java:32]: Implementation looks ok.
Can you move it to `Hooks` class and make a hook out of it, being executed before test suite (`@BeforeAll`)?

File: integration-tests/src/test/java/com/redhat/service/bridge/integration/tests/steps/WebhookSiteSteps.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+                .atMost(Duration.ofMinutes(timeoutMinutes))
+                .pollInterval(Duration.ofSeconds(1))
+                .untilAsserted(() -> assertThat(requests())
+                        .map(request -> request.getContent())
+                        .as("Searching for request containing text: '%s'",
+                                requestTextWithoutPlaceholders)
+                        .anyMatch(requestContent -> requestContent.contains(requestTextWithoutPlaceholders)));
+    }
+
+    private static List<WebhookSiteRequest> requests() {
```
[the remark this answers is not in the document]
Comment (sutaakar) [reply integration-tests/src/test/java/com/redhat/service/bridge/integration/tests/steps/WebhookSiteSteps.java:64]: Same with WebhookSiteRequest class

File: integration-tests/src/test/resources/features/webhookaction.feature (hunk 9 above)
Comment (jomarko) [reply integration-tests/src/test/resources/features/webhookaction.feature]: will do

File: integration-tests/src/test/java/com/redhat/service/bridge/integration/tests/steps/WebhookSiteSteps.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (jomarko) [reply integration-tests/src/test/java/com/redhat/service/bridge/integration/tests/steps/WebhookSiteSteps.java:64]: Sure, will do

File: integration-tests/src/test/java/com/redhat/service/bridge/integration/tests/steps/WebhookSiteSteps.java (hunk 1 above)
Comment (jomarko) [reply integration-tests/src/test/java/com/redhat/service/bridge/integration/tests/steps/WebhookSiteSteps.java]: should be there now

**Review by sutaakar — APPROVED** [review]
Looks good to me now, good work

**Comment by sutaakar** [comment]
```
		Suppressed: io.cucumber.java.InvalidMethodSignatureException: A method annotated with BeforeAll or AfterAll must have one of these signatures:
 * public static void before_or_after_all()
at com.redhat.service.bridge.integration.tests.steps.Hooks.webhookSiteRequestHistoryIsCleared()
```

**Comment by jomarko** [comment]
@sutaakar please review https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/507 first

**Comment by jomarko** [comment]
```
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:203)
	at io.restassured.internal.RequestSpecificationImpl.get(RequestSpecificationImpl.groovy:173)
	at io.restassured.internal.RequestSpecificationImpl.get(RequestSpecificationImpl.groovy)
	at io.restassured.RestAssured.get(RestAssured.java:720)
	at com.redhat.service.bridge.integration.tests.resources.webhook.site.WebhookSiteResource.requests(WebhookSiteResource.java:16)
	at com.redhat.service.bridge.integration.tests.steps.Hooks.webhookSiteRequestHistoryIsCleared(Hooks.java:42)

Error: ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 2.08 s <<< FAILURE! - in RunCucumberTest
Error: ERROR] Cucumber  Time elapsed: 2.065 s  <<< ERROR!
java.net.ConnectException: Connection refused (Connection refused)
```

Let me investigate the failure.

**Review by Srihari1192 — APPROVED** [review]

**Pull request merged**